ledger/participant-integration-api/src/main/scala/platform/apiserver/execution/LedgerTimeAwareCommandExecutor.scala

```diff
@@ -66,39 +78,42 @@ private[apiserver] final class LedgerTimeAwareCommandExecutor(
           else
             contractStore
               .lookupMaximumLedgerTime(usedContractIds)
-              .flatMap { maxUsedTime =>
-                if (maxUsedTime.forall(_ <= commands.commands.ledgerEffectiveTime)) {
-                  Future.successful(Right(cer))
-                } else if (!cer.dependsOnLedgerTime) {
-                  logger.debug(
-                    s"Advancing ledger effective time for the output from ${commands.commands.ledgerEffectiveTime} to $maxUsedTime"
-                  )
-                  Future.successful(Right(advanceOutputTime(cer, maxUsedTime)))
-                } else if (retriesLeft > 0) {
-                  metrics.daml.execution.retry.mark()
-                  logger.debug(
-                    s"Restarting the computation with new ledger effective time $maxUsedTime"
+              .flatMap {
+                case Left(missingContracts) =>
+                  logger.info(
+                    s"Did not find the ledger time for some contracts. This can happen if there is contention on contracts used by the transaction. Missing contracts: ${missingContracts
+                      .mkString(", ")}."
                   )
-                  val advancedCommands = advanceInputTime(commands, maxUsedTime)
-                  loop(advancedCommands, submissionSeed, ledgerConfiguration, retriesLeft - 1)
-                } else {
-                  Future.successful(Left(ErrorCause.LedgerTime(maxRetries)))
-                }
+                  retryOrError(commands)
+
+                case Right(maxUsedTime) =>
+                  if (maxUsedTime.forall(_ <= commands.commands.ledgerEffectiveTime)) {
+                    Future.successful(Right(cer))
+                  } else if (!cer.dependsOnLedgerTime) {
+                    logger.debug(
+                      s"Advancing ledger effective time for the output from ${commands.commands.ledgerEffectiveTime} to $maxUsedTime"
+                    )
+                    Future.successful(Right(advanceOutputTime(cer, maxUsedTime)))
+                  } else {
+                    logger.info(
+                      s"Ledger time was used in the Daml code, but a different ledger effective time $maxUsedTime was determined afterwards. The transaction needs to be rerun with the new ledger effective time."
```

Comment:
If we keep on failing on the time-advancement path, we will eventually print the line above i.e. the intent to `rerun with the new ledger effective time`. But then we might decide not to do it inside of `retryOrError` due to exhausted number of trials.

Reply:
Yeah, I'll delete the last sentence. The rerun logic will then log the next step anyway.